### ISSUE TITLE
Automatic location propagation

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi-azure-nextgen#6]
Similar to the previous Azure provider, we should:

- Make `location` parameter optional for all resources
- If `location` is not specified for a resource, populate it from the resource group's `location`
- If `location` is not specified for a resource group, use the config value `azure-nextgen:location`
- Customize the error message to give a hint about the configuration parameter

**Comment by PawelStadnicki** [comment]
I was about to raise the issue that default location doesn't work for azure-nextgen provider
but first searched current issues and found this one.

This enhancement is highly anticipated as passing additional parameter or putting it in the model is a burden when changing location is unlikely

**Comment by mikhailshilkov** [comment]
Another option is to use transformations. Here is an example for C#:

```cs
static ResourceTransformationResult? AddLocation(ResourceTransformationArgs res)
{
    var locationProp = res.Args.GetType().GetProperty("Location");
    if (locationProp != null)
    {
        locationProp.SetValue(res.Args, (Input<string>)"West Europe");
        return new ResourceTransformationResult(res.Args, res.Options);
    }
    return null;
}

public MyStack() : base(new StackOptions { ResourceTransformations = { AddLocation } })
{
    var resourceGroup = new ResourceGroup("resourceGroup", new ResourceGroupArgs
    {
        ResourceGroupName = "mikhail-rg",
        // no need for this anymore Location = "West Europe",
    });
}
```

Transformations in other languages: https://www.pulumi.com/docs/intro/concepts/programming-model/#transformations

**Comment by PawelStadnicki** [comment]
One more question. Lets assume that instead of a Location I want to "process" resource group name for some reason. To do that I need the resource group actual value but it is a Input<string>. How to get the raw value of it?
I know you cannot do it with Output  but input in this case always will have some string content.
You can do that: 
 (Input<string>) "some text"
but you can't
var name = (Input<string>) as string (or similar).

This is unsafe but should be possible I think but blocked by the framework design?
Is there another way to get resource group raw name of resource instead of Input<string> ?

**Comment by mikhailshilkov** [comment]
Why do you need the raw value? A pattern is to convert an input to output, change it with `Apply`, and assign it elsewhere.

I guess you could get the value with reflection but there's no official way to convert an input to a raw value.

**Comment by PawelStadnicki** [comment]
"Why do you need the raw value?"

This is interesting :) I have many subscriptions. 
In order to provision a resource in a proper subscription I need to pass subscription id inside a provider in CustomResourceOption.

Without transformation I have to pass own CustomResourceOption to each resource I want to use as a parameter which makes my design crap.

In order to get subscription id I'm doing a mapping from resource group names, which makes the transformation approach challenging because with transformation it is only as a Input<string> which requires Apply hence requires return Output which is not acceptable.

However I made a below PoC that seems to work  but I'm not sure if it is robust. 
Definitely it still is not elegant as uses side-effect (setter) instead of method return.

If Input<string> is always a raw string (as my case), not a Task then the code is sync and Apply return immediately?
If yes then I just set the new provider via options setter and return null as a transformation result.

```
static ResourceTransformationResult? MapSubscription(ResourceTransformationArgs res)
        {        
            var resourceGroupName = res.Args.GetType().GetProperty("ResourceGroupName");
            
            if (resourceGroupName != null && resourceGroupName.PropertyType == typeof(Pulumi.Input<string>)) 
            {
                var name = resourceGroupName.GetValue(res.Args) as Input<string>;
                var ignore = name.Apply( rg => { 
                      res.Options.Provider = GetProviderPerSubscriptionId(res.Options.Provider,  rg);
                      return rg; 
                });
            }
            return null;       
        }
```

**Comment by PawelStadnicki** [comment]
When the Transformation is called at all? Can I assume that other resource inputs are "awaited" and RAW values avaiable?

**Comment by mikhailshilkov** [comment]
Transformations are called from the resource constructor, so inputs are not resolved yet https://github.com/pulumi/pulumi/blob/4f0789159596fa540562a0293c8797e49289ecac/sdk/dotnet/Pulumi/Resources/Resource.cs#L152

**Comment by PawelStadnicki** [comment]
Thanks @mikhailshilkov
So my final question is:
If I have the transformation metod and set the provider inside the Apply of Input, that is fine?
In other words: is provider used after all the inputs are resolved or before?

**Comment by mikhailshilkov** [comment]
Here is an example of how to apply basic auto-location and other auto-populated properties with stack transformations: https://github.com/matwilko/Pulumi.AzureNextGen.Ambient

**Comment by imod** [comment]
I think this should to do it for TS:

```
function registerLocation(location: string) {
  pulumi.runtime.registerStackTransformation((args: ResourceTransformationArgs) => {
    if (args.type.startsWith("azure-nextgen:")) {
      if (!args.props["location"]) {
        args.props["location"] = location;
      }
      return {
        props: args.props,
        opts: args.opts,
      };
    }
    return undefined;
  });
}
```

but as `location` is mostly mandatory on the TS types, we have to do it explicitly anyway - or am I wrong?

**Comment by mikhailshilkov** [comment]
@imod Hmm, good point. I guess we'll have to take care of that.

**Comment by mikhailshilkov** [comment]
This is resolved in the next release: `pulumi config set azure-nextgen:location westus`. Alternatively, resources will take location from their resource group. Also, `location` isn't a required property anymore, so transformations can set it too.

**Comment by imod** [comment]
@mikhailshilkov many thanks for implementing this!
Please allow me to give a comment about the resolution: It would be very helpful if you guys could link to the commit/PRs which implement this feature. That way others would have the chance to get a better idea on the insides of the whole pulumi internals and therefore ease the entry to contributions...

**Comment by mikhailshilkov** [comment]
@imod Unfortunately, that code is not open-source at the moment, see #36

The PR is actually linked but you can't see it...

**Comment by imod** [comment]
Uh, I was not aware of that - to be honest: not sure how to deal with that now that I know. We made our decision to use pulumi based on the fact that it actually is open source... we might need to rethink our decision :(

**Comment by mikhailshilkov** [comment]
Well, feel free to comment on that issue and on slack
